mirrord/auth/src/credential_store.rs

```diff
@@ -0,0 +1,85 @@
+use std::{
+    collections::{btree_map::Entry, BTreeMap},
+    fmt::Debug,
+    path::PathBuf,
+    sync::LazyLock,
+};
+
+use kube::{Client, Resource};
+use serde::{Deserialize, Serialize};
+use tokio::fs;
+
+use crate::{
+    credentials::Credentials,
+    error::{AuthenticationError, CertificateStoreError, Result},
+};
+
+static CREDENTIALS_PATH: LazyLock<PathBuf> = LazyLock::new(|| {
+    home::home_dir()
+        .unwrap_or_else(|| PathBuf::from("~"))
+        .join(".mirrord/credentials")
+});
+
+#[derive(Debug, Serialize, Deserialize)]
+pub struct CredentialStore {
+    active: String,
+    common_name: Option<String>,
+    client_credentials: BTreeMap<String, Credentials>,
+}
+
+impl Default for CredentialStore {
+    fn default() -> Self {
+        CredentialStore {
+            active: "default".to_string(),
+            common_name: None,
+            client_credentials: BTreeMap::new(),
+        }
+    }
+}
+
+impl CredentialStore {
+    pub async fn load() -> Result<Self> {
+        let buffer = fs::read(&*CREDENTIALS_PATH)
+            .await
+            .map_err(CertificateStoreError::from)?;
+
+        serde_yaml::from_slice(&buffer)
+            .map_err(CertificateStoreError::from)
+            .map_err(AuthenticationError::from)
+    }
+
+    pub async fn save(&self) -> Result<()> {
+        let buffer = serde_yaml::to_string(&self).map_err(CertificateStoreError::from)?;
+
+        fs::write(&*CREDENTIALS_PATH, &buffer)
+            .await
+            .map_err(CertificateStoreError::from)
+            .map_err(AuthenticationError::from)
+    }
+
+    pub async fn get_or_init<R>(&mut self, client: &Client) -> Result<&mut Credentials>
+    where
+        R: Resource + Clone + Debug,
+        R: for<'de> Deserialize<'de>,
+        R::DynamicType: Default,
+    {
+        let credentials = match self.client_credentials.entry(self.active.clone()) {
+            Entry::Vacant(entry) => entry.insert(Credentials::init()?),
+            Entry::Occupied(entry) => entry.into_mut(),
+        };
+
+        if !credentials.is_ready() {
+            let common_name = self
+                .common_name
+                .clone()
+                .or_else(|| gethostname::gethostname().into_string().ok())
+                .unwrap_or_default();
+
+            credentials
+                .get_client_certificate::<R>(client.clone(), &common_name)
+                .await?;
```

Comment:
The way I understood the requirements, when a user works with different clusters that have different licenses, `mirrord` should identify the correct license automatically by the CN.
In this code, the used certificate is determined by the `active` configuration field, not by which operator mirrord is connecting to.
Is automatic license choosing a requirement, or not @aviramha  / @eyalb181?

Either way - with this code, I think that if a user uses a mirrord operator with one license, and then tries to use a second mirrord operator with a different license, mirrord would try to authenticate with an invalid certificate unless they manually edit the credential store file and set a new `active`. Does that fulfil the requirements? Should users be manually editing that file?
In my opinion it would be better to identify credentials in the store by the operator they are valid for (maybe by the *common name* of the license of that operator?) instead of by the `active` value at the time of their creation. This makes more sense together with automatic credential choosing.

Reply:
Ok, so when we introduce multiple-license support we'll have kind of a backward compatible logic of - if there is no credentials entry for this operator - try "default", and only if that fails - request a new one? Is that the plan, @DmitryDodzin?